pkg/types/portmap.go

```diff
@@ -47,38 +49,75 @@ func (pps PortProtoSet) Equal(other PortProtoSet) bool {
 	return true
 }
 
+// Add increments the reference count for the specified key.
+func (pps PortProtoSet) Add(pp PortProto) bool {
+	return counter.Counter[PortProto](pps).Add(pp)
+}
+
+// Delete decrements the reference count for the specified key.
+func (pps PortProtoSet) Delete(pp PortProto) bool {
+	return counter.Counter[PortProto](pps).Delete(pp)
+}
+
 // NamedPortMultiMap may have multiple entries for a name if multiple PODs
 // define the same name with different values.
 type NamedPortMultiMap interface {
 	// GetNamedPort returns the port number for the named port, if any.
 	GetNamedPort(name string, proto uint8) (uint16, error)
 	// Len returns the number of Name->PortProtoSet mappings known.
 	Len() int
-	Equal(other NamedPortMultiMap) bool
 }
 
-func NewNamedPortMultiMap() namedPortMultiMap {
-	return make(namedPortMultiMap)
+// NamedPortMultiMapUpdater allows for mutation of the NamedPortMultiMap, and is
+// intended for ipcache internal usage.
+type NamedPortMultiMapUpdater interface {
+	NamedPortMultiMap
+	Update(old, new NamedPortMap) (namedPortChanged bool)
+}
+
+func NewNamedPortMultiMapUpdater() NamedPortMultiMapUpdater {
+	return &namedPortMultiMap{
+		m: make(map[string]PortProtoSet),
+	}
 }
 
-type namedPortMultiMap map[string]PortProtoSet
+// Implements both NamedPortMultiMap and its Updater variant. All accesses to it
+// must be protected by its RW mutex.
+type namedPortMultiMap struct {
+	lock.RWMutex
+	m map[string]PortProtoSet
+}
 
-// Equal returns true if the NamedPortMultiMaps are equal.
-func (npm namedPortMultiMap) Equal(other NamedPortMultiMap) bool {
-	o, ok := other.(namedPortMultiMap)
-	if !ok || len(npm) != len(o) {
-		return false
+func (npm *namedPortMultiMap) Len() int {
+	npm.RLock()
+	defer npm.RUnlock()
+	return len(npm.m)
+}
+
+// Update applies potential changes in named ports, and returns whether there were any.
+func (npm *namedPortMultiMap) Update(old, new NamedPortMap) (namedPortsChanged bool) {
+	npm.Lock()
+	defer npm.Unlock()
+	// The order is important here. Increment the refcount first, and then
+	// decrement it again for old ports, so that we don't hit zero if there are
+	// no changes.
+	for name, port := range new {
+		c, ok := npm.m[name]
+		if !ok {
+			c = make(PortProtoSet)
+			npm.m[name] = c
+		}
+		if c.Add(port) {
+			namedPortsChanged = true
+		}
 	}
-	for name, ports := range npm {
-		if otherPorts, exists := o[name]; !exists || !ports.Equal(otherPorts) {
-			return false
+	for name, port := range old {
+		namedPortsChanged = npm.m[name].Delete(port) || namedPortsChanged
```

Comment:
tiny nit: I _think_ we only need to check the len if `npm.m[name].Delete(port)` returned true. I don't think there is any other condition when it would be zero.

```go
	if npm.m[name].Delete(port) {
		namedPortsChanged = true
		if len(npm.m[name]) == 0 {
			delete(npm.m, name)
		}
	}
```

(As a bonus point, it would bring back the aestethic symmetry with Add, which uses an `if`-statement instead of `||` to set `namedPortsChanged`)

Reply:
I agree, looks better and passes tests, changed. Thanks!